File: headergenerator.py
import re

def extract_functions_constants_classes(file_path):
    functions = []
    constants = []
    classes = []
    includes = set()

    with open(file_path, 'r') as file:
        lines = file.readlines()

        in_comment_block = False
        current_function = ""
        current_class = ""

        for line in lines:
            # Check for comment blocks
            if '/*' in line:
                in_comment_block = True
            if '*/' in line:
                in_comment_block = False
                continue
            
            if not in_comment_block:
                # Extract functions

                function_match = re.match(r'^\s*([\w:<>]+)\s+([\w:<>]+)\s*\((.*)\)\s*', line) or re.match(r'^\s*([\w:<>]+)\s+([\w:<>]+)\s*\((.*)\)\s*{', line)
                if function_match:
                    return_type = function_match.group(1)
                    function_name = function_match.group(2)
                    arguments = function_match.group(3)
                    functions.append((return_type, function_name, arguments))
                
                # Extract constants 
                constant_match = re.match(r'^\s*const\s+([\w:<>]+)\s+([\w]+)\s*=\s*([^;]+);', line) or re.match(r'^\s*constexpr\s+([\w:<>]+)\s+([\w]+)\s*=\s*([^;]+);', line)
                if constant_match:
                    constant_type = constant_match.group(1)
                    constant_name = constant_match.group(2)
                    constant_value = constant_match.group(3)
                    constants.append((constant_type, constant_name, constant_value))
                
                # Extract classes
                class_match = re.match(r'^\s*class\s+([\w:<>]+)', line)
                if class_match:
                    class_name = class_match.group(1)
                    classes.append(class_name)

                # Extract includes
                include_match = re.match(r'^\s*#include\s+<(.*)>', line) or re.match(r'^\s*#include\s+"(.*)"', line) 
                if include_match:
                    includes.add(include_match.group(1))

    return functions, constants, classes, includes

def write_to_header(header_path, functions, constants, classes, includes):
    with open(header_path, 'w') as header_file:
        for include in includes:
            header_file.write(f'#include "{include}"\n')


        header_file.write('\n')

        for constant in constants:
            header_file.write(f'constexpr {constant[0]} {constant[1]} = {constant[2]};\n')

        if constants:
            header_file.write('\n')

        for function in functions:
            header_file.write(f'{function[0]} {function[1]}({function[2]});\n')

        if functions:
            header_file.write('\n')

        for class_name in classes:
            header_file.write(f'class {class_name};\n')


def generate_header(cpp_file_path, header_file_path):

    functions, constants, classes, includes = extract_functions_constants_classes(cpp_file_path)
    write_to_header(header_file_path, functions, constants, classes, includes)

File: headergenerator.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <set>
#include <regex>


void extract_functions_constants_classes(const std::string& file_path, 
std::vector<std::pair<std::string, std::string>>& functions,std::vector
<std::pair<std::string, std::string>>& constants,
std::vector<std::string>& classes,
std::set<std::string>& includes){
std::ifstream file(file_path);
if (!file.is_open()) {
    std::cerr << "Failed to open file: " << file_path << std::endl;
    return;
}

std::string line;
bool in_comment_block = false;
std::string current_function = "";
std::string current_class = "";

std::regex function_regex("^\\s*([\\w:<>]+)\\s+([\\w:<>]+)\\s*\\((.*)\\)\\s*");
std::regex function_regex2("^\\s*([\\w:<>]+)\\s+([\\w:<>]+)\\s*\\((.*)\\)\\s*\\{");
std::regex constant_regex("^\\s*const\\s+([\\w:<>]+)\\s+([\\w]+)\\s*=\\s*([^;]+);");
std::regex constant_regex2("^\\s*constexpr\\s+([\\w:<>]+)\\s+([\\w]+)\\s*=\\s*([^;]+);");
std::regex class_regex("^\\s*class\\s+([\\w:<>]+)");
std::regex include_regex("^\\s*#include\\s+<(.*)>");
std::regex include_regex2("^\\s*#include\\s+\"(.*)\"");
    
    while (getline(file, line)) {
        // Check for comment blocks
        if (line.find("/*") != std::string::npos) {
            in_comment_block = true;
        }
        if (line.find("*/") != std::string::npos) {
            in_comment_block = false;
            continue;
        }
        
        if (!in_comment_block) {
            // Extract functions
            std::smatch match;
            if (regex_match(line, match, function_regex) || regex_match(line, match, function_regex2)) {
                std::string return_type = match[1];
                std::string function_name = match[2];
                std::string arguments = match[3];
                functions.emplace_back(return_type, function_name + "("+ arguments +")");
            }
            
            // Extract constants 
            if (regex_match(line, match, constant_regex) || regex_match(line, match, constant_regex2)) {
                std::string constant_type = match[1];
                std::string constant_name = match[2];
                std::string constant_value = match[3];
                constants.emplace_back(constant_type, constant_name);
            }
            
            // Extract classes
            if (regex_match(line, match, class_regex)) {
                std::string class_name = match[1];
                classes.push_back(class_name);
            }
            
            // Extract includes
            if (regex_match(line, match, include_regex) || regex_match(line, match, include_regex2)) {
                includes.insert(match[1]);
            }
        }
    }
}

void write_to_header(const std::string& header_path,
                     const std::vector<std::pair<std::string, std::string>>& functions,
                     const std::vector<std::pair<std::string, std::string>>& constants,
                     const std::vector<std::string>& classes,
                     const std::set<std::string>& includes) {
    std::ofstream header_file(header_path);

    if (!header_file.is_open()) {
        std::cerr << "Failed to open file for writing: " << header_path << std::endl;
        return;
    }

    for (const auto& include : includes) {
        header_file << "#include \"" << include << "\"\n";
    }

    header_file << "\n";

    for (const auto& constant : constants) {
        header_file << "constexpr " << constant.first << " " << constant.second << ";\n";
    }

    if (!constants.empty()) {
        header_file << "\n";
    }

    for (const auto& function : functions) {
        header_file << function.first << " " << function.second << ";\n";
    }

    if (!functions.empty()) {
        header_file << "\n";
    }

    for (const auto& class_name : classes) {
        header_file << "class " << class_name << ";\n";
    }
}

void generate_cpp_header(std::string cpp_file_path, std::string header_file_path) {
    std::vector<std::pair<std::string, std::string>> functions;
    std::vector<std::pair<std::string, std::string>> constants;
    std::vector<std::string> classes;
    std::set<std::string> includes;

    extract_functions_constants_classes(cpp_file_path, functions, constants, classes, includes);
    write_to_header(header_file_path, functions, constants, classes, includes);
}



